### PR TITLE
[BEAM-788] Implement GroupAlsoByWindowsEvaluator Directly

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

Invoke the ReduceFnRunner methods directly rather than by constructing a
GroupAlsoByWindowDoFn and a ParDoEvaluator. Instead, construct a
ReduceFnRunner, ReduceFn, and invoke the methods directly.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @kennknowles

**Comment by amitsela** [comment]
@tgroh shouldn't this be at the `runner-core` level, instead of having each runner implement this ?

**Comment by tgroh** [comment]
`ReduceFnRunner` is the shared `runner-core` abstraction for executing (in the most general case) a `GroupAlsoByWindow`. We currently invoke the `ReduceFnRunner` via the `GroupAlsoByWindowViaWindowSetDoFn`; however, that isn't really a `DoFn`, as it requires direct access to `StateInternals`, `TimerInternals`, and `WindowingInternals` - this PR merely moves the behavior of `GroupAlsoByWindowViaWindowSetDoFn` into the `GroupAlsoByWindowsEvaluator` to enable removal of `OldDoFn`

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/GroupAlsoByWindowEvaluatorFactory.java [hunk 1]
```diff
@@ -97,65 +129,215 @@ public GroupAlsoByWindowEvaluator(
...
+    /**
+     * Returns an {@code Iterable<WindowedValue<InputT>>} that only contains non-late input
+     * elements.
+     */
+    public Iterable<WindowedValue<V>> dropExpiredWindows(
+        final K key, Iterable<WindowedValue<V>> elements, TimerInternals timerInternals) {
+      return FluentIterable.from(elements)
+          .transformAndConcat(new ExplodeWindowsFn<V>())
+          .filter(
+              new ElementWindowIsExpiredPred<V>(
```
Comment (kennknowles) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/GroupAlsoByWindowEvaluatorFactory.java]: I kind of feel like the functional style is causing a lot of boilerplate.

**Comment by kennknowles** [comment]
+1 to the idea of basically inlining this. I like the idea of having something high level in `runners-core` but there is so little here that I'm not sure what would be factored out. In fact, use of `GroupAlsoByWindowsViaWindowSetDoFn` forces a particular materialization of timers, while each runner might just like to use their own representation.

To be clear, you've also reproduced the late data dropping, which might be nice to have a way to re-use.

And it can be further shrunk (not needed for this PR). `ReduceFnRunner.onTimer(TimerData)` actually doesn't need the timer data, it just needs to fire EOW panes, fire trigger panes, and GC. The `TimerData` is just a hint about which windows to process.

**Pull request closed without merging**